benchmarks/QueryBench.php

```diff
@@ -0,0 +1,103 @@
+<?php
+
+namespace Amp\Mysql\Bench;
+
+use Amp\Mysql\Connection;
+use Amp\Mysql\ConnectionPool;
+use Amp\Mysql\Internal\ConnectionConfig;
+use Amp\Mysql\ResultSet;
+use PhpBench\Benchmark\Metadata\Annotations\BeforeClassMethods;
+use PhpBench\Benchmark\Metadata\Annotations\BeforeMethods;
+use PhpBench\Benchmark\Metadata\Annotations\AfterMethods;
+use PhpBench\Benchmark\Metadata\Annotations\Iterations;
+use PhpBench\Benchmark\Metadata\Annotations\Revs;
+use PhpBench\Benchmark\Metadata\Annotations\Warmup;
+use PhpBench\Benchmark\Metadata\Annotations\OutputTimeUnit;
+use function Amp\Promise\wait;
+use function Amp\Promise\all;
+use function Amp\call;
+
+/**
+ * @BeforeMethods({"init"})
+ * @AfterMethods({"onAfterMethods"})
+ * @Iterations(1)
+ * @Revs(10)
+ * @Warmup(1)
+ * @OutputTimeUnit("milliseconds", precision=5)
+ */
+class QueryBench extends AbstractBench
+{
+    /** @var  ConnectionPool */
+    protected $connectionPool;
+    /** @var  Connection */
+    protected $connection;
+    /** @var  \PDO */
+    protected $pdoConnection;
+    protected $maxQueries = 10;
+    protected $poolLimit = 10;
+
+    public function init()
+    {
+        $config = ConnectionConfig::parseConnectionString("host=$this->host;user=$this->user;pass=$this->pass");
+        $this->connectionPool = new ConnectionPool($config, $this->poolLimit);
+        $connectionPromise = Connection::connect($config);
+        $this->connection = wait($connectionPromise);
+        $this->pdoConnection = new \PDO("mysql:host=$this->host;", $this->user, $this->pass);
+    }
+
+    public function onAfterMethods()
+    {
+        $this->connectionPool->close();
+        $this->connection->close();
+    }
+
+    public function benchPdoQueries()
+    {
+        foreach (range(1, $this->maxQueries) as $ii) {
+            $resultSet = $this->pdoConnection->query("SELECT $ii");
+            $resultSet->fetchAll();
+        }
+    }
+
+    public function benchSyncQueries()
+    {
+        wait(call(function() {
+            $connection = $this->connection;
+            foreach (range(1, $this->maxQueries) as $i) {
+                /** @var ResultSet $resultSet */
+                $resultSet = yield $connection->query("SELECT $i");
+                yield $resultSet->advance();
+            }
+        }));
+    }
+
+    public function benchAsyncQueries()
+    {
+        wait(call(function() {
+            $connection = $this->connection;
+            /** @var ResultSet[] $resultSets */
+            $resultSets = yield all(array_map(function ($i) use ($connection) {
+                return $connection->query("SELECT $i");
+            }, range(1, $this->maxQueries)));
+            yield all(array_map(function ($resultSet) {
+                /** @var ResultSet $resultSet */
+                return $resultSet->advance();
+            }, $resultSets));
```

Comment:
You can leave of the `all` if you want, it will implicitly be applied if you yield an array of promises, but that's personal preference, just a hint if you didn't know.

Reply:
Thank you. Not sure, but probably a previous version of amp had required `all` here. Cool, now much clearer.

Reply:
Yes, 1.x didn't allow yielding array of promises.